### DISCUSSION TITLE
Calling caf::disposable::dispose causes the monitor callback function to block indefinitely

### DISCUSSION THREAD

**Discussion by phil-skillwon** [actor-framework/actor-framework#1977] [Q&A]
May I ask a question about using the new version of the `monitor` member function? In my code, I’m storing the `caf::disposable` handle returned by `monitor` as a member variable inside the actor (e.g. `caf::disposable m_dispHandle`). I’ve written a callback function that captures `this` so it can access `m_dispHandle` inside the callback, and then it calls `m_dispHandle.dispose()`. However, I’m noticing that the entire program blocks at that point and the callback never returns. 

Does this mean I shouldn’t call `caf::disposable::dispose` from inside the callback?

Below is roughly what my C++ code looks like:

```cpp
m_dispHandle = this->monitor(proxyActor, [this, proxy = proxyActor](const caf::error& reason){
    // ...
    // ...
    // ...
    this->m_dispHandle.dispose(); // The entire program blocks here indefinitely
    // ...
});
```

The CAF documentation for `dispose` says:

```cpp
/// Disposes the resource. Calling `dispose()` on a disposed resource is a
/// no-op.
void dispose() {
  if (pimpl_) {
    pimpl_->dispose();
    pimpl_ = nullptr;
  }
}
```

My understanding is that calling this member function should be safe whether the monitored actor is still alive or not. So is it incorrect to access the disposable handle from within the callback? Or am I missing something here?

**Comment by Neverlord** [comment]
@phil-skillwon sorry for the late response, this got probably buried since it has been opened in the holiday season.

The disposable points to the action itself that runs the callback to `monitor`. So disposing it in the lambda means we are disposing the action while it's running. Looks like we are currently holding onto a mutex in the `monitor_action` while running it. So calling `dispose` deadlocks because of that. Of course we should not hold onto the mutex while running the callback, that's a bug.

**Reply by Neverlord** [reply]
Will be fixed in CAF 1.2: https://github.com/actor-framework/actor-framework/issues/2169
